##PDS PRACTICAL 3

**NAME:VARAD VIJAY AMBEKAR**

**DIV: SY-A**

**ROLL NO.: A-02**

**Aim = Data Wrangling with pandas: Manipulate and transform data to suit the analysis     
needs.**

**1 . What is data wrangling? Explain the difference between data wrangling and data preprocessing.**

Data Wrangling
Data wrangling means cleaning, organizing, and transforming raw data into a useful format for analysis.

It includes tasks like:

-Removing missing values

-Correcting wrong data

-Changing data types

-Renaming columns

-Combining multiple datasets

-Formatting data properly

Difference between Data Wrangling and Data Preprocessing

Data Wrangling:

-Focuses on cleaning and organizing raw data

-Includes handling missing values, duplicates, formatting

Data Preprocessing:

-Focuses on preparing data for machine learning or analysis

-Includes normalization, scaling, encoding

**2 . What is the importance of exploratory data analysis (EDA) in Pandas?**
EDA means **Exploratory Data Analysis**.It is used to understand the data before performing analysis or building models.

Importance of EDA in Pandas:
-Helps to understand the structure of data

-Finds missing values

-Detects outliers

-Checks data types

-Shows patterns and relationships

-Helps in better decision making

-Improves data quality

Common Pandas functions used in EDA
head()

tail()

info()

describe()

value_counts()

**3.What are the main differences between Series, DataFrame, and Panel?**

1.Series

-One-dimensional data structure

-Like a single column

-Stores values with index

Example:

import pandas as pd

s = pd.Series([10, 20, 30])

2.DataFrame

-Two-dimensional data structure

-Rows and columns

-Like a table or Excel sheet

Example:

df = pd.DataFrame(
  {
    'Name': ['A', 'B'],
    'Marks': [80, 90]
})

3.Panel

-Three-dimensional data structure

-Collection of multiple DataFrames

**4 . Explain the use of:**

a. isnull()

b. notnull()

c. dropna()

d. fillna()

**1 . isnull()**


Used to check missing (null) values in data.


Returns True for missing values.

Example:

df.isnull()


b) notnull()

-Used to check non-missing values.

-Returns True where value is present.

Example:

df.notnull()



c) dropna()

-Used to remove rows or columns containing missing values.

Example:

df.dropna()


d) fillna()

-Used to replace missing values with some value.

Example:

df.fillna(0)

**5 . What is column transformation and why is it important in data analysis?
Column Transformation**

Column transformation means changing the values or format of a column to make it useful for analysis.

Examples:

-Converting text to numbers

-Changing uppercase to lowercase

-Scaling values

-Creating new columns from old ones

Example:

df['Age'] = df['Age'] + 1

In [ ]:
import pandas as pd

df = pd.read_csv("/content/train (1).csv")
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
age_column = df["Age"]
print(age_column.head())

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64


In [ ]:
df["Source"] = "Titanic Dataset"
print(df[["Name", "Source"]].head())

                                                Name           Source
0                            Braund, Mr. Owen Harris  Titanic Dataset
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  Titanic Dataset
2                             Heikkinen, Miss. Laina  Titanic Dataset
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  Titanic Dataset
4                           Allen, Mr. William Henry  Titanic Dataset


In [ ]:
sorted_df = df.sort_values(by="Fare", ascending=False)
print(sorted_df[["Name", "Fare"]].head())

                                   Name      Fare
679  Cardeza, Mr. Thomas Drake Martinez  512.3292
258                    Ward, Miss. Anna  512.3292
737              Lesurer, Mr. Gustave J  512.3292
88           Fortune, Miss. Mabel Helen  263.0000
438                   Fortune, Mr. Mark  263.0000


In [ ]:
df_renamed = df.rename(columns={
    "Pclass": "Passenger_Class",
    "SibSp": "Siblings_Spouses",
    "Parch": "Parents_Children"
})
print(df_renamed.columns.tolist())

['PassengerId', 'Survived', 'Passenger_Class', 'Name', 'Sex', 'Age', 'Siblings_Spouses', 'Parents_Children', 'Ticket', 'Fare', 'Cabin', 'Embarked', 'Source']


In [ ]:
print(df.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


In [ ]:
grouped = df.groupby("Pclass")[["Fare", "Age"]].mean()
print(grouped)

             Fare        Age
Pclass                      
1       84.154687  38.233441
2       20.662183  29.877630
3       13.675550  25.140620


In [ ]:
extra_info = pd.DataFrame({
    "PassengerId": [1, 2, 3, 4, 5],
    "Nationality": ["British", "American", "Irish", "French", "German"]
})

merged_df = pd.merge(df, extra_info, on="PassengerId", how="left")
print(merged_df[["PassengerId", "Name", "Nationality"]].head())

   PassengerId                                               Name Nationality
0            1                            Braund, Mr. Owen Harris     British
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...    American
2            3                             Heikkinen, Miss. Laina       Irish
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)      French
4            5                           Allen, Mr. William Henry      German


In [ ]:
df_dropped_cols = df.drop(columns=["Cabin", "Ticket", "Source"])
print("After dropping columns:", df_dropped_cols.shape)

After dropping columns: (891, 10)


In [ ]:
df_dropped_rows = df_dropped_cols.dropna(subset=["Age"])
print("After dropping rows with missing Age:", df_dropped_rows.shape)

After dropping rows with missing Age: (714, 10)


In [ ]:
def age_group(age):
    if age < 18:
        return "Child"
    elif age < 60:
        return "Adult"
    else:
        return "Senior"

df["Age_Group"] = df["Age"].apply(age_group)
print(df[["Name", "Age", "Age_Group"]].head(10))


                                                Name   Age Age_Group
0                            Braund, Mr. Owen Harris  22.0     Adult
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0     Adult
2                             Heikkinen, Miss. Laina  26.0     Adult
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0     Adult
4                           Allen, Mr. William Henry  35.0     Adult
5                                   Moran, Mr. James   NaN    Senior
6                            McCarthy, Mr. Timothy J  54.0     Adult
7                     Palsson, Master. Gosta Leonard   2.0     Child
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  27.0     Adult
9                Nasser, Mrs. Nicholas (Adele Achem)  14.0     Child


In this practical, we performed Data Wrangling using the pandas library in Python on the Titanic dataset (train.csv) containing 891 rows
and 12 columns such as PassengerId, Survived, Pclass, Age, Fare, Sex, and more. Basic Operations (Tasks 1–5) covered the foundational
skills of pandas — selecting columns, adding new columns with constant values, sorting data, renaming columns, and generating
summary statistics using describe(). These operations help in understanding and organizing the raw dataset. Moderate Operations
(Tasks 6–10) involved more analytical tasks — filtering rows using multiple conditions (e.g., female passengers who survived), grouping
data by passenger class to find average Fare and Age, merging two DataFrames on a common key, dropping unwanted rows and
columns to clean the data, and applying custom functions to create new derived columns like Age Group (Child / Adult / Senior).
Through these 10 tasks, we learned how pandas provides powerful and flexible tools to clean, transform, filter, and analyze structured
data efficiently. These operations form the core of any real-world data analysis or machine learning pipeline, making pandas an
essential skill for any Data Science practitioner.